satpy/composites/viirs.py

```diff
@@ -85,18 +86,37 @@ def __init__(self, *args, **kwargs):
         self.dem_sds = kwargs.pop("dem_sds", "averaged elevation")
         super(ReflectanceCorrector, self).__init__(*args, **kwargs)
 
-    def __call__(self, datasets, **info):
-        refl_data, sensor_aa, sensor_za, solar_aa, solar_za = datasets
+    def __call__(self, datasets, optional_datasets, **info):
+        if not optional_datasets or len(optional_datasets) != 4:
+            vis = self.check_areas([datasets[0]])[0]
+            sensor_aa, sensor_za, solar_aa, solar_za = self.get_angles(vis)
+        else:
+            vis, sensor_aa, sensor_za, solar_aa, solar_za = self.check_areas(
+                datasets + optional_datasets)
+            # get the dask array underneath
+            sensor_aa = sensor_aa.data
+            sensor_za = sensor_za.data
+            solar_aa = solar_aa.data
+            solar_za = solar_za.data
+        # angles must be xarrays
+        sensor_aa = xr.DataArray(sensor_aa, dims=['y', 'x'])
+        sensor_za = xr.DataArray(sensor_za, dims=['y', 'x'])
+        solar_aa = xr.DataArray(solar_aa, dims=['y', 'x'])
+        solar_za = xr.DataArray(solar_za, dims=['y', 'x'])
+        refl_data = datasets[0]
         if refl_data.attrs.get("rayleigh_corrected"):
             return refl_data
-
         if os.path.isfile(self.dem_file):
             LOG.debug("Loading CREFL averaged elevation information from: %s",
                       self.dem_file)
             from netCDF4 import Dataset as NCDataset
             # HDF4 file, NetCDF library needs to be compiled with HDF4 support
             nc = NCDataset(self.dem_file, "r")
-            avg_elevation = nc.variables[self.dem_sds][:]
+            # average elevation is stored as a 16-bit signed integer but with
+            # scale factor 1 and offset 0, convert it to float here
+            avg_elevation = nc.variables[self.dem_sds][:].astype(np.float)
```

Comment:
I think in the long run we'll want this to be dask arrays, but for now let's not worry about it. It probably isn't a performance issue anyway.